rasterio/_io.pyx

```diff
@@ -890,9 +895,22 @@ cdef class DatasetWriterBase(DatasetReaderBase):
 
         if mode == 'w':
 
-            # Delete existing file, create.
-            if os.path.exists(path):
-                os.unlink(path)
+            # Delete dataset if it already exists.
```

Comment:
This doesn't use `rasterio.exists()` and `rasterio.delete()` because of some circular import issues, which could be solved by dropping those two functions in `rasterio._io`, but having them in `rasterio._manage` alongside other data management tooling feels a little more organized.  This implementation can take a few shortcuts too.